### PR TITLE
Binder annotations

### PR DESCRIPTION
**Pull request by 333fred**
Annotate more of the binder.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs [hunk 1]
```diff
@@ -129,7 +129,7 @@ private BoundExpression RewriteDeconstruction(BoundTupleExpression left, Convers
             for (int i = 0; i < leftTargets.Count; i++)
             {
                 BoundExpression resultPart;
-                if (leftTargets[i].HasNestedVariables)
+                if (leftTargets[i].NestedVariables is object)
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: Could we annotate HasNestedVariables somehow to make NestedVariables null-checked?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: Not yet. We're going to be adding an API for this in C# 9, `MemberNotNull`, but there's nothing there for it today.

File: src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs [hunk 2]
```diff
@@ -695,9 +699,9 @@ private static string ExtractDeconstructResultElementName(BoundExpression expres
         private BoundBadExpression MissingDeconstruct(BoundExpression receiver, SyntaxNode rightSyntax, int numParameters, DiagnosticBag diagnostics,
                                     out ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders, BoundExpression childNode)
         {
-            if (!receiver.Type.IsErrorType())
+            if (!(receiver.Type?.IsErrorType() ?? false))
             {
-                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, rightSyntax, receiver.Type, numParameters);
+                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, rightSyntax, receiver.Type!, numParameters);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs]: this feels odd, we explicitly do enter this block when 'receiver.Type' is null. Don't we?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs]: I misplaced a paren, thanks.

**Comment by 333fred** [comment]
Verified that this builds cleanly on a master merge locally.

**Pull request merged**